src/qibojit/custom_operators/backends.py

```diff
@@ -384,3 +383,252 @@ def swap_pieces(self, piece0, piece1, new_global, nlocal):
     def measure_frequencies(self, frequencies, probs, nshots, nqubits, seed=1234, nthreads=None):
         raise NotImplementedError("`measure_frequencies` method is not "
                                   "implemented for GPU.")
+
+
+class CuQuantumBackend(CupyBackend): # pragma: no cover
+    # CI does not test for GPU
+
+    def __init__(self):
+        super(CuQuantumBackend, self).__init__()
+        import cuquantum # pylint: disable=import-error
+        from cuquantum import custatevec as cusv # pylint: disable=import-error
+        self.cuquantum = cuquantum 
+        self.cusv = cusv
+        self.name = "cuquantum"
+
+    def get_cuda_type(self, dtype='complex64'):
+        if dtype == 'complex128':
+            return self.cuquantum.cudaDataType.CUDA_C_64F, self.cuquantum.ComputeType.COMPUTE_64F
+        elif dtype == 'complex64':
+            return self.cuquantum.cudaDataType.CUDA_C_32F, self.cuquantum.ComputeType.COMPUTE_32F
+        else:
+            raise TypeError("Type can be either complex64 or complex128")
+
+    def one_qubit_base(self, state, nqubits, target, kernel, qubits=None, gate=None):
+        # initialize cuStateVec library
+        handle = self.cusv.create()
+
+        ntarget = 1
+        target = nqubits - target - 1
+        target = self.np.asarray([target], dtype = self.np.int32)
+        ncontrols = len(qubits) - 1 if qubits is not None else 0
+        controls = self.np.asarray([i for i in qubits.get() if i != target], dtype = self.np.int32)
+        adjoint = 0
+        # implement x, y, z and z_pow kernel since cuquantum only has apply_matrix
+        # TODO: find a better way to implement this
```

Comment:
Another way to implement this (not sure if better) is the following:

Change the `apply_x`, `apply_y`, `apply_z` methods in `JITCustomBackend` to pass the gate matrix:
```Python
def apply_x(self, state, nqubits, targets, qubits=None):
    return self.engine.one_qubit_base(state, nqubits, *targets, "apply_x", qubits, self.matrices.X)
```
and ignore the matrix in the cupy backend by chaging the `if gate is None:` condition in `one_qubit_base` to
```Python
if kernel in ("apply_x", "apply_y", "apply_z"):
```
I think for the numba backend the gate will be ignored as it is.

Reply:
Ok, I can give this a try. Initially I coded the matrices directly in the `JITCustomBackend` methods, then I move them to the `CuQuantumBackend`.